homeassistant/components/yolink/__init__.py

```diff
@@ -49,12 +58,83 @@
-        self,
-        hass: HomeAssistant,
+        self, hass: HomeAssistant, yl_client: YoLinkClient, yl_mqtt_client: MqttClient
     ) -> None:
-        """Init YoLink DataUpdateCoordinator."""
-        super().__init__(hass, _LOGGER, name=DOMAIN)
+        """Init YoLink DataUpdateCoordinator.
+
+        fetch state every 30 minutes base on yolink device heartbeat interval
+        """
+        super().__init__(
+            hass, _LOGGER, name=DOMAIN, update_interval=timedelta(minutes=30)
+        )
+        self._client = yl_client
+        self._mqtt_client = yl_mqtt_client
+        self.data = {}
+
+    async def init_coordinator(self):
+        """Init coordinator."""
+        try:
+            async with async_timeout.timeout(10):
+                home_info = await self._client.get_general_info()
+
+                def on_message_callback(message: tuple[str, BRDP]):
+                    """On message callback."""
+                    data = message[1]
+                    if data.event is None:
+                        return
+                    event_param = data.event.split(".")
+                    event_type = event_param[len(event_param) - 1]
+                    if event_type not in (
+                        "Report",
+                        "Alert",
+                        "StatusChange",
+                        "getState",
+                    ):
+                        return
+                    resolved_state = data.data
+                    if resolved_state is None:
+                        return
+                    self.data[message[0]] = resolved_state
+                    self.async_set_updated_data(self.data)
+
+                await self._mqtt_client.init_home_connection(
+                    home_info.data["id"], on_message_callback
+                )
+            async with async_timeout.timeout(10):
+                device_response = await self._client.get_auth_devices()
+                yl_devices: list[YoLinkDevice] = []
+                for device_info in device_response.data[ATTR_DEVICE]:
+                    yl_device = YoLinkDevice(device_info, self._client)
+                    yl_devices.append(yl_device)
+                self.yl_devices = yl_devices
+        except YoLinkAuthFailError as yl_auth_err:
+            raise ConfigEntryAuthFailed from yl_auth_err
+        except (YoLinkClientError, asyncio.TimeoutError) as err:
+            raise ConfigEntryNotReady from err
+
+    async def fetch_device_state(self, device: YoLinkDevice):
+        """Fetch Device State."""
+        async with async_timeout.timeout(20):
+            device_state_resp = await device.fetch_state_with_api()
+            if ATTR_DEVICE_STATE in device_state_resp.data:
+                self.data[device.device_id] = device_state_resp.data[ATTR_DEVICE_STATE]
+
+    async def _async_update_data(self) -> dict:
+        try:
+            for yl_device in self.yl_devices:
+                await self.fetch_device_state(yl_device)
```

Comment:
Does this need to happen sequentially or can we gather the work with `asyncio.gather`?